### DISCUSSION TITLE
[feature Request] Support multipart and resumable upload

### DISCUSSION THREAD

**Discussion by wipphj** [filebrowser/filebrowser#1182] [Ideas]
**Is your feature request related to a problem? Please describe.**
As a user, i use cloudflare as a cdn provider,But cloudflare  has  100mb upload limit,so i cannot  upload large files

**Describe the solution you'd like**
Use fragmented upload methods to circumvent cloudflare restrictions

**Describe alternatives you've considered**
use the library such as resumable.js ，Of course, the server must be adjusted accordingly

**Additional context**

**Comment by o1egl** [comment]
I've made some research and found https://tus.io/ more mature and suitable for this project to implement. It also supports partial uploads as part of the spec.

**Reply by wipphj** [reply]
it's a good idea 😀

**Reply by t3pfaffe** [reply]
Seconded on it being a good idea. Would then allow for use of CDN's like Cloudflare who have limitations on the maximum sizes of POSTs.

**Reply by kinghat** [reply]
there is also: https://github.com/23/resumable.js

**Reply by Toiongo** [reply]
Would really appreciate such support as well.

**Reply by ohpls** [reply]
Yeah that would be very good to have as feature.

**Reply by HackMEAny** [reply]
Yes, will this implement soon ???

**Reply by august-isaque** [reply]
Can't wait to have this feature implemented in the project 👍 :)

**Reply by cainnz** [reply]
+1 please let us know when it is implemented

**Comment by wait-what** [comment]
+1 on this. Cloudflare's 100MB free upload limit is a pain, so if the file were to be uploaded in chunks, this could be bypassed.

**Comment by Tusharkumar0012** [comment]
Same problem. Much need d feature.
One way for uploading large files is to temporarily turn off dns proxy at cloudflare and then upload large files and proxy again .
But it also has risk of exposing ip.

**Comment by TobiasGoerke** [comment]
I've just opened a [pull request](https://github.com/filebrowser/filebrowser/pull/2145) which integrates tus.io into the filebrowser. Currently, this lets you upload files in 20mb chunks and provides resumable uploads, too.
You may either build the image yourself (see PR for a Dockerfile) or use the dockerhub image I've provided for testing: _tobiasgoerke/filebrowser:tus_. 
I hope this may help you, too. I'm looking forward to reading your feedback. Let me know in case you find any bugs.

**Reply by august-isaque** [reply]
Does it provides a easier way to lock the file while uploading? Preventing any other application that might be running/watching the same directory to read/moving it?

I would like to have something like the winscp that when you upload a file, it first create a file with a .temp extension, and when the upload is done, it removes the .temp...

**Reply by TobiasGoerke** [reply]
In my implementation, all chunks are uploaded to a tmp folder within the user directory. After all chunks are uploaded, they are combined, moved to their destination, and the tmp files are deleted. This, the original file isn't touched until the upload is completed.

**Reply by Tusharkumar0012** [reply]
> I've just opened a [pull request](https://github.com/filebrowser/filebrowser/pull/2145) which integrates tus.io into the filebrowser. Currently, this lets you upload files in 20mb chunks and provides resumable uploads, too.
> You may either build the image yourself (see PR for a Dockerfile) or use the dockerhub image I've provided for testing: _tobiasgoerke/filebrowser:tus_. 
> I hope this may help you, too. I'm looking forward to reading your feedback. Let me know in case you find any bugs. 

Your work is appreciated 👍. Will test it out when get time. If the pull request is merged will we be able to update existing docker images ?

**Reply by gonzalomorenorovetto** [reply]
@TobiasGoerke 
Hey, just installed your docker image. It works great for me, except for spaces in the directories or filenames
They are replaced by %20 when merging the chunks and moving the files, which doesn't work. It think spaces should be replaced with "\ " but I don't know how to do that.
Here's an example of my log:
"Could not move file from /srv/.tmp_upload/f53a7dea176f64b95b5bb2a3e69e047d to /srv/DKMS/2022/LATAM%20DKMS/CAM%201/A054_05281650_C001.mov: rename /srv/.tmp_upload/f53a7dea176f64b95b5bb2a3e69e047d /srv/DKMS/2022/LATAM%20DKMS/CAM%201/A054_05281650_C001.mov: no such file or directory"

I think the full route should be something like "/srv/DKMS/2022/LATAM\ DKMS/CAM\ 1/A054_05281650_C001.mov"

If I upload the same files, but to a directory with no spaces on the route, it works fine

**Reply by TobiasGoerke** [reply]
> @TobiasGoerke Hey, just installed your docker image. It works great for me, except for spaces in the directories or filenames They are replaced by %20 when merging the chunks and moving the files, which doesn't work. It think spaces should be replaced with "\ " but I don't know how to do that. Here's an example of my log: "Could not move file from /srv/.tmp_upload/f53a7dea176f64b95b5bb2a3e69e047d to /srv/DKMS/2022/LATAM%20DKMS/CAM%201/A054_05281650_C001.mov: rename /srv/.tmp_upload/f53a7dea176f64b95b5bb2a3e69e047d /srv/DKMS/2022/LATAM%20DKMS/CAM%201/A054_05281650_C001.mov: no such file or directory"
> 
> I think the full route should be something like "/srv/DKMS/2022/LATAM\ DKMS/CAM\ 1/A054_05281650_C001.mov"
> 
> If I upload the same files, but to a directory with no spaces on the route, it works fine

This issue has been fixed a couple of months ago, I just didn't rebuild the image. I did so now, so you can pull tobiasgoerke/filebrowser:tus.

Please keep in mind you shouldn't run unverified images (like this) for anything other than testing. Instread, it'd be much better to pull this image from the official filebrowser repo. This, however, would require my PR to get merged. As the contributors have not responded for a long time, feel free to push the PR.

**Reply by gonzalomorenorovetto** [reply]
Hey, thanks!

**Reply by jfmadridc** [reply]
> > @TobiasGoerke Hey, just installed your docker image. It works great for me, except for spaces in the directories or filenames They are replaced by %20 when merging the chunks and moving the files, which doesn't work. It think spaces should be replaced with "\ " but I don't know how to do that. Here's an example of my log: "Could not move file from /srv/.tmp_upload/f53a7dea176f64b95b5bb2a3e69e047d to /srv/DKMS/2022/LATAM%20DKMS/CAM%201/A054_05281650_C001.mov: rename /srv/.tmp_upload/f53a7dea176f64b95b5bb2a3e69e047d /srv/DKMS/2022/LATAM%20DKMS/CAM%201/A054_05281650_C001.mov: no such file or directory"
> > I think the full route should be something like "/srv/DKMS/2022/LATAM\ DKMS/CAM\ 1/A054_05281650_C001.mov"
> > If I upload the same files, but to a directory with no spaces on the route, it works fine
> 
> This issue has been fixed a couple of months ago, I just didn't rebuild the image. I did so now, so you can pull tobiasgoerke/filebrowser:tus.
> 
> Please keep in mind you shouldn't run unverified images (like this) for anything other than testing. Instread, it'd be much better to pull this image from the official filebrowser repo. This, however, would require my PR to get merged. As the contributors have not responded for a long time, feel free to push the PR.

@TobiasGoerke any chance of an arm version/tag? would really appreciated
keep up the good work

**Reply by TobiasGoerke** [reply]
Thanks, @jfmadridc! I'm relucant providing images myself as these wouldn't be verified. Instead, I've been waiting a long time for this PR to get merged so that filebrowser would directly provide the images. But I'm not receiving a response from the maintaners and honestly don't know if this project is still alive.
